### ISSUE TITLE
Improving foreground alignment

### ISSUE THREAD

**Issue by m43** [facebookresearch/map-anything#14]
Thanks again for the great work! It's very convenient that one can condition the model with known camera intrinsics and extrinsics, this improves the results a lot in some cases. Do you have any tips on how to let the model better align the foreground in the examples below, or in what ways to improve/retrain the model to have this work out better?

<table>
  <tr>
    <th>4D Dress (pose-free)</th>
    <th>4D Dress (known poses)</th>
  </tr>
  <tr>
    <td><video src="https://github.com/user-attachments/assets/f9da0464-a1d2-4e24-bb00-baa170b0410a" controls></video></td>
    <td><video src="https://github.com/user-attachments/assets/c4761aef-262c-4cb5-8959-f7846d1956c6" controls></video></td>
  </tr>
</table>

<table>
  <tr>
    <th>DexYCB (pose-free)</th>
    <th>DexYCB (known poses)</th>
  </tr>
  <tr>
    <td><video src="https://github.com/user-attachments/assets/5bceca4f-425c-4a04-855e-2f0474b303ed" controls></video></td>
    <td><video src="https://github.com/user-attachments/assets/b225a5d1-c681-4618-bb0d-4d55f0f0542a" controls></video></td>
  </tr>
</table>

Reproducibility snippet for DexYCB:
```py
import os

os.environ["PYTORCH_CUDA_ALLOC_CONF"] = "expandable_segments:True"  # Optional, for better memory efficiency

import rerun as rr
import numpy as np
import torch
from tqdm import tqdm

from huggingface_hub import hf_hub_download

from mapanything.models import MapAnything
from mapanything.utils.image import preprocess_inputs, find_closest_aspect_ratio
from mapanything.utils.cropping import crop_resize_if_necessary
from mapanything.models import MapAnything
from mapanything.utils.geometry import depthmap_to_world_frame
from mapanything.utils.image import load_images
from mapanything.utils.viz import (
    predictions_to_glb,
    script_add_rerun_args,
)
from uniception.models.encoders.image_normalizations import IMAGE_NORMALIZATION_DICT


def log_data_to_rerun(
        image, depthmap, pose, intrinsics, pts3d, mask, base_name, pts_name, viz_mask=None
):
    """Log visualization data to Rerun"""
    # Log camera info and loaded data
    height, width = image.shape[0], image.shape[1]
    rr.log(
        base_name,
        rr.Transform3D(
            translation=pose[:3, 3],
            mat3x3=pose[:3, :3],
        ),
    )
    rr.log(
        f"{base_name}/pinhole",
        rr.Pinhole(
            image_from_camera=intrinsics,
            height=height,
            width=width,
            camera_xyz=rr.ViewCoordinates.RDF,
            image_plane_distance=1.0,
        ),
    )
    rr.log(
        f"{base_name}/pinhole/rgb",
        rr.Image(image),
    )
    rr.log(
        f"{base_name}/pinhole/depth",
        rr.DepthImage(depthmap),
    )
    if viz_mask is not None:
        rr.log(
            f"{base_name}/pinhole/mask",
            rr.SegmentationImage(viz_mask.astype(int)),
        )

    # Log points in 3D
    filtered_pts = pts3d[mask]
    filtered_pts_col = image[mask]

    rr.log(
        pts_name,
        rr.Points3D(
            positions=filtered_pts.reshape(-1, 3),
            colors=filtered_pts_col.reshape(-1, 3),
        ),
    )


def main():
    # Create MapAnything model
    device = "cuda" if torch.cuda.is_available() else "cpu"
    model = MapAnything.from_pretrained("facebook/map-anything").to(device)

    # Download MVTracker demo sample from Hugging Face Hub
    sample_path = hf_hub_download(
        repo_id="ethz-vlg/mvtracker",
        filename="data_sample.npz",
        token=os.getenv("HF_TOKEN"),
        repo_type="model",
    )
    sample = np.load(sample_path)
    rgbs = torch.from_numpy(sample["rgbs"])
    depths = torch.from_numpy(sample["depths"]).float()
    intrs = torch.from_numpy(sample["intrs"]).float()
    extrs = torch.from_numpy(sample["extrs"]).float()
    extrs[..., :3, 3] /= 6  # Revert sample to metric scale (MVTracker sample is scaled up by 6x)
    n_views, n_frames, _, H, W = rgbs.shape

    # extrs is world2cam, inverse is cam2world needed for MapAnything
    extrs_square = torch.eye(4, device=extrs.device)[None].repeat(n_views, n_frames, 1, 1)
    extrs_square[:, :, :3, :] = extrs
    extrs_inv = torch.inverse(extrs_square.float()).type(extrs.dtype)

    assert rgbs.shape == (n_views, n_frames, 3, H, W)
    assert depths.shape == (n_views, n_frames, 1, H, W)
    assert intrs.shape == (n_views, n_frames, 3, 3)
    assert extrs.shape == (n_views, n_frames, 3, 4)
    assert extrs_inv.shape == (n_views, n_frames, 4, 4)

    # Initialize rerun
    rr.init(f"3dpt", recording_id="v0.16")
    rr.connect_tcp()

    for t in tqdm(range(n_frames), desc=f"MapAnything inference on {n_frames} frames"):
        input_views = []
        for v in range(n_views):
            input_views.append({
                "img": rgbs[v, t].permute(1, 2, 0).contiguous().numpy(),
                "intrinsics": intrs[v, t].numpy().astype(np.float32),
                "camera_poses": extrs_inv[v, t].numpy().astype(np.float32),
                "is_metric_scale": torch.tensor([True], device=device),  # (1,)
            })
        views = preprocess_inputs(
            input_views,
            resize_mode="fixed_mapping",
            norm_type="dinov2",
            patch_size=14,
            resolution_set=518,
            verbose=True,
        )
        predictions = model.infer(
            views,  # Input views
            memory_efficient_inference=False,  # Trades off speed for more views (up to 2000 views on 140 GB)
            use_amp=True,  # Use mixed precision inference (recommended)
            amp_dtype="bf16",  # bf16 inference (recommended; falls back to fp16 if bf16 not supported)
            apply_mask=True,  # Apply masking to dense geometry outputs
            mask_edges=True,  # Remove edge artifacts by using normals and depth
            apply_confidence_mask=False,  # Filter low-confidence regions
            confidence_percentile=10,  # Remove bottom 10 percentile confidence pixels
        )

        # Log to rerun for debugging
        for view_idx, pred in enumerate(predictions):
            # Extract data from predictions
            depthmap_torch = pred["depth_z"][0].squeeze(-1)  # (H, W)
            intrinsics_torch = pred["intrinsics"][0]  # (3, 3)
            camera_pose_torch = pred["camera_poses"][0]  # (4, 4)

            # Compute new pts3d using depth, intrinsics, and camera pose
            pts3d_computed, valid_mask = depthmap_to_world_frame(
                depthmap_torch, intrinsics_torch, camera_pose_torch
            )

            # Convert to numpy arrays
            mask = pred["mask"][0].squeeze(-1).cpu().numpy().astype(bool)
            mask = mask & valid_mask.cpu().numpy()  # Combine with valid depth mask
            pts3d_np = pts3d_computed.cpu().numpy()
            image_np = pred["img_no_norm"][0].cpu().numpy()

            rr.set_time_seconds("frame", t / 12)
            log_data_to_rerun(
                image=image_np,
                depthmap=depthmap_torch.cpu().numpy(),
                pose=camera_pose_torch.cpu().numpy(),
                intrinsics=intrinsics_torch.cpu().numpy(),
                pts3d=pts3d_np,
                mask=mask,
                base_name=f"mapanything/view_{view_idx}",
                pts_name=f"mapanything/pointcloud_view_{view_idx}",
                viz_mask=mask,
            )

    print("Done")


if __name__ == '__main__':
    main()
```

**Comment by Nik-V9** [comment]
Thanks for trying it out! The first failure case is quite funny 😂

Are the poses metric scale?

I would recommend trying out `ignore_pose_scale_inputs=True` to see if that helps.

Also MapAnything wasn’t trained with a lot of dynamic datasets apart from a few synthetic ones. We have seen that finetuning it particularly on video datasets or the ones typically used for scene flow like PointOdyssey improves performance on non rigid objects.

**Comment by m43** [comment]
> Are the poses metric scale?

Yes, they are metric. I tried with `ignore_pose_scale_inputs=True`, but it looks about the same and didn't help with the foreground.

> MapAnything wasn’t trained with a lot of dynamic datasets apart from a few synthetic ones

Okk, got it! I actually run MapAnything per-frame because of that, so the videos above are just per-frame reconstructions, that's why it is a little jittery over time. Maybe the training data didn't have static multi-view captures of humans or small foreground objects?

**Comment by Nik-V9** [comment]
Yes, we don’t really have any static multi view captures of humans in the training data.

I think this new dataset which was used to train Pi3 should capture this use case: https://arxiv.org/abs/2509.12201

**Comment by Nik-V9** [comment]
Btw could you also share the 4Dress data?

**Comment by m43** [comment]
> Btw could you also share the 4D-Dress data?

It’s unfortunately under [this license](https://4d-dress.ait.ethz.ch), which explicitly prohibits redistribution. :( To access it, you’d need to fill out the request form yourself and be granted the license. Once you have the license, I could email you my post-processed pickle file under the same license terms

**Comment by Nik-V9** [comment]
> It’s unfortunately under [this license](https://4d-dress.ait.ethz.ch), which explicitly prohibits redistribution. :( To access it, you’d need to fill out the request form yourself and be granted the license. Once you have the license, I could email you my post-processed pickle file under the same license terms

Sounds good, I'll do this and send you an email! At CMU, we can try out the to be released model I mentioned in our chat on this :)

**Comment by m43** [comment]
That'd be amazing! :)

**Comment by Nik-V9** [comment]
Please try the new V1.1 checkpoints or Any4D and let me know if things improve! :)
